plonky2/src/hash/keccak.rs

```diff
@@ -61,23 +59,28 @@ impl<F: RichField> PlonkyPermutation<F> for KeccakPermutation<F> {
     }
 
     fn permute(&mut self) {
-        let mut state_bytes = vec![0u8; SPONGE_WIDTH * size_of::<u64>()];
-        for i in 0..SPONGE_WIDTH {
+        let mut state_bytes = [0u8; SPONGE_WIDTH * size_of::<u64>()];
+        for (i, x) in self.state.iter().enumerate() {
             state_bytes[i * size_of::<u64>()..(i + 1) * size_of::<u64>()]
-                .copy_from_slice(&self.state[i].to_canonical_u64().to_le_bytes());
+                .copy_from_slice(&x.to_canonical_u64().to_le_bytes());
         }
 
-        let hash_onion = core::iter::repeat_with(|| {
-            let output = keccak(state_bytes.clone()).to_fixed_bytes();
-            state_bytes = output.to_vec();
-            output
+        let hash_onion = (0..).scan(keccak(state_bytes), |state, _| {
```

Comment:
Why do you use `scan` here and, `repeat_with` elsewhere?

(This one could be `repeat_with` and `take`.  Or the other version could turn into a `scan`.)

`successors` might also work, and would probably be the cleanest, if it does.